components/gitpod-db/src/typeorm/code-sync-resource-db.ts

```diff
@@ -70,38 +70,43 @@ export class CodeSyncResourceDB {
         });
     }
 
-    async insert(userId: string, kind: ServerResource, doInsert: (rev: string, oldRev?: string) => Promise<void>, params?: CodeSyncInsertOptions): Promise<string | undefined> {
+    async deleteResource(userId: string, kind: ServerResource, rev: string): Promise<void> {
+        const connection = await this.typeORM.getConnection();
+        await connection.transaction(async manager => {
+            await manager.createQueryBuilder()
+                    .delete()
+                    .from(DBCodeSyncResource)
+                    .where("userId = :userId AND kind = :kind AND rev = :rev", { userId, kind, rev: rev })
+                    .execute();
+        });
+    }
+
+    async insert(userId: string, kind: ServerResource, doInsert: (rev: string, oldRevs: string[]) => Promise<void>, params?: CodeSyncInsertOptions): Promise<string | undefined> {
         const connection = await this.typeORM.getConnection();
         return await connection.transaction(async manager => {
             let latest: DBCodeSyncResource | undefined;
-            let toUpdated: DBCodeSyncResource | undefined;
+            let toDeleted: DBCodeSyncResource[] = [];
             if (params?.revLimit) {
                 const resources = await this.doGetResources(manager, userId, kind);
                 latest = resources[0];
                 if (resources.length >= params.revLimit) {
-                    toUpdated = resources[resources.length - 1];
+                    // delete + 1 to insert instead of update
+                    toDeleted = resources.splice(params?.revLimit - 1);
                 }
             } else {
                 latest = await this.doGetResource(manager, userId, kind, 'latest');
             }
+            // TODO(hw): why we use `latest?.rev !== params.latestRev` ? what's latestRev?
```

Comment:
latestRev is the `etag` of the resource, if the `etag` send from the client is different from the server then it means it got updated already by some other client so we return 412 error, here's an example of how [they mock it](https://github.com/microsoft/vscode/blob/98c97abb8a33a34a3b82434c91ef93fe5ac819fc/src/vs/platform/userDataSync/test/common/userDataSyncClient.ts#L250-L252) in a test

Reply:
@mustard-mh that's a good reference, i used this test impl as a blueprint while implementing our code sync endpoing

Reply:
Got it! 🙏